presto-main/src/main/java/com/facebook/presto/server/QueryResource.java

```diff
@@ -100,13 +109,34 @@ public void getAllQueryInfo(
             proxyResponse(servletRequest, asyncResponse, xForwardedProto, uriInfo);
             return;
         }
-        QueryState expectedState = stateFilter == null ? null : QueryState.valueOf(stateFilter.toUpperCase(Locale.ENGLISH));
-        ImmutableList.Builder<BasicQueryInfo> builder = new ImmutableList.Builder<>();
-        for (BasicQueryInfo queryInfo : dispatchManager.getQueries()) {
-            if (stateFilter == null || queryInfo.getState() == expectedState) {
-                builder.add(queryInfo);
-            }
+
+        int limit = firstNonNull(limitFilter, Integer.MAX_VALUE);
+        if (limit <= 0) {
+            throw new WebApplicationException(Response
+                    .status(BAD_REQUEST)
+                    .type(MediaType.TEXT_PLAIN)
+                    .entity(format("Parameter 'limit' for getAllQueryInfo must be positive. Got %d.", limit))
+                    .build());
+        }
+
+        List<BasicQueryInfo> queries = new ArrayList<>(dispatchManager.getQueries());
+
+        // Filter list by the query state (if specified).
+        if (stateFilter != null) {
+            QueryState expectedState = QueryState.valueOf(stateFilter.toUpperCase(Locale.ENGLISH));
+            queries.removeIf(item -> item.getState() == expectedState);
         }
+
+        // If limit is smaller than number of queries, then ensure that the more recent items are at the front.
+        if (limit < queries.size()) {
+            queries.sort(Comparator.comparing(item -> item.getQueryStats().getCreateTime()));
```

Comment:
Rather than simply most recent, I think we should prefer running queries, then queued queries, then all other non-completed queries, then all failed queries, then all successfully completed queries.  We can tweak this ordering, but the basic idea is, we don't want to filter out long running queries from this list that are still running, we should prefer currently active queries first since those likely have more interest than completed queries; likewise, failed queries are likely to be of more interest than completed queries.

It could look something like this:

```suggestion
            queries.sort(
                Ordering.<BasicQueryInfo>from((o1, o2) -> Boolean.compare(o1.getState() == RUNNING, o2.getState() == RUNNING))
                        .compound((o1, o2) -> Boolean.compare(o1.getState() == QUEUED, o2.getState() == QUEUED))
                        .compound((o1, o2) -> Boolean.compare(!o1.getState().isDone(), !o2.getState().isDone()))
                        .compound((o1, o2) -> Boolean.compare(o1.getState() == FAILED, o2.getState() == FAILED))
                        .compound(comparing(item -> item.getQueryStats().getCreateTime())));
```